justfile

```diff
@@ -0,0 +1,18 @@
+default:
+  @just --list
+
+# Cargo build everything.
+build:
+  cargo build --workspace --all-targets --all-features
+
+# Cargo check everything.
+check:
+  cargo check --workspace --all-targets --all-features
+
+# Lint everything.
+lint:
+  cargo clippy --workspace --all-targets --all-features -- --deny warnings
+
+# Run the formatter
+format:
+  cargo +nightly fmt --all
```

Comment:
Is it from Fedimint or do you actually expect people to run nightly formatter?

Reply:
@dpc we have to run the nightly formatter because the stable one butchers the codebase.

Reply:
Oh, OK. In Fedimint the nix dev shell sets up toolchain with stable compiler, but nightly rustfmt. But motivation was mostly some sweet unstable formatting options.

Reply:
Yep, same here. Pretty nice that Nix can do this automatically.

Reply:
Can achieve the same thing in guix @stevenroose?

Reply:
@tcharding it's doubtful because "nightly" is not a well-defined thing. When you use nightly compilers in Nix then you'll get different results on different invocatons, something that Nix allows because it's "pragmatic" while Guix does not because it takes reproducibility seriously.

Reply:
*cough, cough*

I will have you know that Nix takes reproducibility seriously, and there's nothing "ill-defined" about nightly.

In Fedimint `rustfmt` comes from nightly via [`fenix` project](https://github.com/nix-community/fenix) and is locked via `flake.lock` [which includes checking its cryptographic checksum ](https://github.com/fedimint/fedimint/blob/87d68d8356c1d56c27c91de4fc70ee0952b07b13/flake.lock#L98).

Fedimint artifacts produced via Nix are byte-reproducible.

You can verify it by going to [some `master` branch build](https://github.com/fedimint/fedimint/actions/runs/5957088701/job/16159128912), taking the git commit (here: `87d68d8356c1d56c27c91de4fc70ee0952b07b13`) running:

```
nix build 'github:fedimint/fedimint?rev=87d68d8356c1d56c27c91de4fc70ee0952b07b13#fedimint-pkgs' && sha256sum result/bin/*
```

waiting approximately forever, and then comparing the checksums with ones printed by the CI:

![image](https://github.com/rust-bitcoin/rust-bitcoin/assets/9209/2d3b3141-12a9-40fc-8be1-22c56c122f19)


But yes, AFAICT Nix is more pragmatic. We can run Steam, systemd and use language with the usual number of parenthesis.